lib/index.js

```diff
@@ -0,0 +1,507 @@
+module.exports = Clocker
+
+var path = require('path')
+
+var level = require('level')
+var mkdirp = require('mkdirp')
+var parseTime = require('parse-messy-time')
+var strftime = require('strftime')
+var through2 = require('through2')
+
+var util = require('./util')
+var getDate = util.getDate
+
+function Clocker (options) {
+  this.dir = options.dir
+
+  // Create data directory if not existing
+  if (!this.dir) {
+    var HOME = process.env.HOME || process.env.USERPROFILE
+    this.dir = path.join(HOME, '.clocker')
+  }
+  mkdirp.sync(this.dir)
+
+  this.db = level(path.join(this.dir, 'db'), { valueEncoding: 'json' })
+}
+
+Clocker.KEY_FORMAT = 'time!%F %T'
+
+Clocker.RESERVED_DATA_ENTRIES = [
+  'start',
+  'end'
+]
+
+Clocker.getDataObject = function (row) {
+  var obj = {}
+  obj.key = toStamp(row.key)
+  obj.data = row.value || {}
+
+  obj.start = new Date(row.key.split('!')[1])
+  obj.end = (obj.data.end ? new Date(obj.data.end) : 'NOW')
+
+  var milliseconds = (obj.data.end ? obj.end : new Date()) - obj.start
+  var seconds = Math.floor(milliseconds / 1000)
+  obj.elapsed = seconds
+
+  if (obj.data.end) {
+    delete obj.data.end
+  }
+
+  return obj
+}
+
+Clocker.formatElapsed = function (elapsed) {
+  var n = elapsed
+  var hh = pad(Math.floor(n / 60 / 60), 2)
+  var mm = pad(Math.floor(n / 60 % 60), 2)
+  var ss = pad(Math.floor(n % 60), 2)
+  return [ hh, mm, ss ].join(':')
+}
+
+Clocker.prototype.close = function (cb) {
+  this.db.close(cb)
+}
+
+Clocker.prototype.start = function (data, date, cb) {
+  // Normalize input
+  // (1) data argument
+  if (typeof data === 'function') {
+    cb = data
+    date = null
+    data = null
+  } else if (typeof data === 'string' || data instanceof Date) {
+    // missing data argument
+    cb = date
+    date = data
+    data = null
+  }
+  // (2) date argument
+  if (typeof date === 'function') {
+    cb = date
+    date = null
+  } else if (typeof date === 'string') {
+    date = parseTime(date)
+  }
```

Comment:
I am not really sure, but wouldn't it be better to change the function signature to `cb, data=null, date=null` to avoid all the normalization stuff? `cb, data, date` should be also good enough since you do the `data = data || new Date()` stuff a few lines below.

This is just one example of this kind of normalization. I've found them a lot in the code and it feels a little bit strange to me. 🙊

Reply:
Yeah, I know, it just follows node.js' design pattern of *callback as last argument*. Maybe there is a shiny npm module that handles all the switch/case of possibly optional arguments?

Reply:
I think https://github.com/ORESoftware/pragmatik could be used for that. But it seems to be a lot of overhead, too. So maybe it is good enough for now.

Perhaps we can investigate if a promise-based solution will work fine in the future. 🙂